clippy_lints/src/bytecount.rs

```diff
@@ -0,0 +1,80 @@
+use consts::{constant, Constant};
+use rustc_const_math::ConstInt;
+use rustc::hir::*;
+use rustc::lint::*;
+use utils::{match_type, paths, snippet, span_lint_and_sugg, walk_ptrs_ty};
+
+/// **What it does:** Checks for naive byte counts
+///
+/// **Why is this bad?** The [`bytecount`](https://crates.io/crates/bytecount)
+/// crate has methods to count your bytes faster, especially for large slices.
+///
+/// **Known problems:** If you have predominantly small slices, the
+/// `bytecount::count(..)` method may actually be slower. However, if you can
+/// ensure that less than 2³²-1 matches arise, the `naive_count_32(..)` can be
+/// faster in those cases.
+///
+/// **Example:**
+///
+/// ```rust
+/// &my_data.filter(|&x| x == 0u8).count() // use bytecount::count instead
+/// ```
+declare_lint! {
+    pub NAIVE_BYTECOUNT,
+    Warn,
```

Comment:
My motivation for writing this lint is that it was only recently brought to my attention that rust-bio used a naive count which they had even optimized on 64-bit x86 systems by using a 32bit count.

I wish more folks knew about bytecount, it's a small dependency considering it can give us awesome speedups.